lib/internal/modules/esm/fetch.js

```diff
@@ -0,0 +1,321 @@
+      }
+      /**
+       * @type {CacheEntry}
+       */
+      const entry = {
+        resolvedHREF: parsed.href,
+        headers: {
+          'content-type': res.headers['content-type']
+        },
+        body: new Promise((f, r) => {
+          const buffers = [];
+          let size = 0;
+          let bodyStream = res;
+          let onError;
+          if (res.headers['content-encoding'] === 'br') {
+            bodyStream = createBrotliDecompress();
+            onError = function onError(error) {
+              bodyStream.close();
+              dispose();
+              reject(error);
+              r(error);
+            };
+            res.on('error', onError);
+            res.pipe(bodyStream);
+          } else if (res.headers['content-encoding'] === 'gzip' ||
+            res.headers['content-encoding'] === 'deflate') {
+            bodyStream = createUnzip();
+            onError = function onError(error) {
+              bodyStream.close();
+              dispose();
+              reject(error);
+              r(error);
+            };
+            res.on('error', onError);
+            res.pipe(bodyStream);
+          } else {
+            onError = function onError(error) {
+              dispose();
+              reject(error);
+              r(error);
+            };
+          }
+          bodyStream.on('error', onError);
+          bodyStream.on('data', (d) => {
+            ArrayPrototypePush(buffers, d);
+            size += d.length;
+          });
+          bodyStream.on('end', () => {
+            const body = entry.body = /** @type {Buffer} */(
+              BufferConcat(buffers, size)
+            );
+            f(body);
+          });
+        }),
+      };
+      cacheForGET.set(parsed.href, entry);
+      fulfill(entry);
+    });
+  });
+  cacheForGET.set(parsed.href, result);
+  return result;
+}
+
+/**
+ * Returns if an address has local status by if it is going to a local
+ * interface or is an address resolved by DNS to be a local interface
+ * @param {string} hostname url.hostname to test
+ * @returns {Promise<boolean>}
+ */
+async function isLocalAddress(hostname) {
+  try {
+    if (StringPrototypeStartsWith(hostname, '[') &&
+        StringPrototypeEndsWith(hostname, ']')) {
+      hostname = StringPrototypeSlice(hostname, 1, -1);
+    }
+    const addr = await dnsLookup(hostname, { verbatim: true });
+    if (addr.family === 4) {
+      return StringPrototypeStartsWith(addr.address, '127.');
+    } else if (addr.family === 6) {
+      return normalizeIPv6(addr.address) ===
+        '0000:0000:0000:0000:0000:0000:0000:0001';
+    }
+  } catch {}
+  return false;
+}
+
+/**
+ * Used to expand interface addresses to a canonical form for comparisons
+ * E.G. comparing ::1, 0::1, etc. to 0000:0000:0000:0000:0000:0000:0000:0001
+ * Due to IPv6 having multiple ways to represent a single address we need to
+ * normalize
+ * @param {string} str
+ * @returns {string}
+ */
+function normalizeIPv6(str) {
```

Comment:
Alternatively, you can make use of the `net.SocketAddress` API here, which performs normalization already.

```js
const n1 = new net.SocketAddress({ address: '::1', family: 'ipv6' });
const n2 = new net.SocketAddress({ address: '0000:0000:0000:0000:0000:0000:0000:0001', famliy: 'ipv6' });

console.log(n1.address == n2.address); // true
```

Less room for error by reusing logic.

Reply:
I switched `isLocalAddress()` to using `net.BlockList()`, and `list.check()` did not play nice with this `normalizeIPv6()`. I removed `normalizeIPv6()`, but I'm not sure it is actually necessary (or if `BlockList::check()` already handles normalisation). The tests pass without it, so if normalisation is required, I should probably add a test case to ensure it's in place.

Reply:
JS wrapper for blocklist already normalizes to a socketaddress